API additions for PcapNG custom blocks

martinling opened the pull request:
I've been working on saving additional data to PcapNG files using custom blocks.

Custom blocks are a feature of the PcapNG format that allow it to be independently extended. Blocks defined by different organisations are identified by [Private Enterprise Numbers](https://www.iana.org/assignments/enterprise-numbers/) assigned by IANA.

There are two distinct custom block types: copiable and non-copiable. Copiable blocks are those whose meaning is self-contained, and which may be freely copied to the output of a program which does not understand their content. Non-copiable blocks are those whose meaning is dependent on any aspect of the format or content of the rest of the PcapNG file. Such blocks may not be copied without understanding their content, as they may become invalid or change meaning.

This PR adds the API features necessary to work effectively with custom blocks, including encoding and decoding payload formats in ways that reuse features of the existing PcapNG protocol state (e.g. endianness and timestamp formats).

The key elements are:

New block type constants.
```rust
/// Custom block type, copiable
pub const CUSTOM_BLOCK_COPIABLE: u32 = 0x00000BAD;
/// Custom block type, non-copiable
pub const CUSTOM_BLOCK_NON_COPIABLE: u32 = 0x40000BAD;
```

The `CustomBlock` type.
```rust
pub struct CustomBlock<'a, const COPIABLE: bool> {
    /// Private Enterprise Number of the entity which defined this block.
    pub pen: u32,
    /// Payload of this block.
    pub payload: Cow<'a, [u8]>,
}
```
New `Block` variants.
```rust
/// Custom block, copiable
CustomCopiable(CustomBlock<'a, true>),
/// Custom block, non-copiable
CustomNonCopiable(CustomBlock<'a, false>),
```
The `PcapNgCustom` trait, which can be implemented for user-defined payload types.
```rust
/// Common interface for custom block payloads
pub trait PcapNgCustom<'a> {
    /// Try to parse this payload from a slice.
    fn from_slice<B: ByteOrder>(state: &PcapNgState, slice: &'a [u8]) -> Result<Option<Self>, PcapError>;

    /// Write this payload into a writer.
    fn write_to<B: ByteOrder, W: Write>(&self, state: &PcapNgState, writer: &mut W) -> Result<usize, PcapError>;
}
```
The `new` and `interpret` methods on `CustomBlock`, which simplify writing and reading custom block formats that implement the above trait.
```rust
/// Create a new custom block from a PEN, a payload and a [`PcapNgState`].
pub fn new<T: PcapNgCustom<'a>>(pen: u32, payload: &T, state: &PcapNgState) -> Result<Self, PcapError>

/// Converts this block's payload into a type that implements [`PcapNgCustom`].
pub fn interpret<'b: 'a, T: PcapNgCustom<'a>>(&'b self, state: &PcapNgState) -> Result<Option<T>, PcapError>;
```
Another important addition is the `PcapNgReader::next_block_and_state` method, which returns a reference to the current `PcapNgState` at the same time as returning the next block.
```rust
/// Returns the next [`Block`] and the current [`PcapNgState`].
pub fn next_block_and_state(&mut self) -> Option<Result<(Block, &PcapNgState), PcapError>> {
```
This is necessary to make it possible to use the `CustomBlock::interpret` method without angering the borrow checker. 

Adding this last method required some internal refactoring of `ReadBuffer::parse_with`, to make it accept an internal `Parser` trait rather than a closure. This was necessary to break the parsing work into separate steps which have different borrowing requirements.

Additionally, new `section` and `interfaces` methods are added to `PcapNgState`, and the `encode_timestamp` and `decode_timestamp` are made public, to support their use in custom blocks.

For an example of usage, see the [custom block definition](https://github.com/greatscottgadgets/packetry/blob/ec775804943b6340b18f55e3cc488eedba1d0c96/src/file.rs#L557-L609), [writing](https://github.com/greatscottgadgets/packetry/blob/ec775804943b6340b18f55e3cc488eedba1d0c96/src/file.rs#L522-L531) and [reading](https://github.com/greatscottgadgets/packetry/blob/ec775804943b6340b18f55e3cc488eedba1d0c96/src/file.rs#L370-L396) in https://github.com/greatscottgadgets/packetry/pull/227.

amorenoz reviewed (commented):
Hi @martinling , I was also looking for adding support to custom blocks (for [retis](https://github.com/retis-org/retis/)) and was very happy to see you had kicked off the work. Thank you!

I did try your series and have 1 comment  from an API / usage PoV.

amorenoz commented on src/pcapng/blocks/custom.rs:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    /// Create a new custom block from a PEN, a payload and a [`PcapNgState`].
+    pub fn new<T: PcapNgCustom<'a>>(pen: u32, payload: &T, state: &PcapNgState) -> Result<Self, PcapError> {
```
>Hmm. If your block doesn't depend on any aspect of the PcapNgState, does that imply it's a copiable block?

Indeed, my blocks are state-less and therefore copiable. Reading the spec, it does seem the relationship with other blocks (i.e: parsing state) is the main rationale behind the flag:

> A common reason for not copying a Custom Block or Custom Option is because it depends on other blocks or options in some way that would invalidate the custom data if the other blocks/options were removed or re-ordered

I ignore if there are any other reasons why a block might not be copiable but it seems a reasonable assumption.

martinling commented:
> Also, looking at the reading part , would it make sense to add functions to `PcapNgParser`, `PcapNgReader` (and internally, to `Block`) to allow users to provide a mapping between custom `PEN` numbers and `PcapNgCustom` generators so that dispatching the core library can take care of partially reading the block (just not the payload) and dispatching it to the right parser?

I thought about this, but it's not straightforward.

The key problem is that the standard doesn't specify a 1-1 mapping between PENs and block formats. The PEN only identifies the _organisation_ that defined the block. And there's no additional field to specify _which_ format from that organisation. There's just a PEN and then raw bytes.

So it's up to each organisation to decide how to distinguish between its own formats, e.g. by specifying that their PEN will always be followed by a type field. But because there's no standard for how to do that, the library can't take care of that part of dispatch for you.

amorenoz commented:
If a user wants to create more than one format for a single PEN, it could register a function that would figure out and do another level of dispatching if desired. The function could just return a `CustomBlock` internally pointing to whatever custom type.

But for simple use-cases where a user wants to create a custom block, write it and read it, forcing them to reimplement the `PcapNgParser` seems a bit too much IMHO.

amorenoz commented:
> I've had a look at that prototype.
> 
> As I said, I'm keen on the part where for writing, you should be able to just construct a block without needing a state, and have it converted statefully to bytes within the writer as per any other block.
> 

> However, the read side here is quite ugly. Implementing `PcapNgCustom` now requires several extra methods, plus there's the runtime overheads of the extra indirection and vtables.
> 

It's true, the API is not more ugly.


> And I don't really see what it's gaining us in exchange. You end up with an opaque pointer with no explicit indication of the payload type. If there's more than one possibility, you'd need to sequentially try downcasting to each of them. From the caller's point of view, that doesn't really seem any easier or tidier than calling `CustomBlock::interpret::<MyCustomPayload>` for each type as with the current version.
> 
> I suppose that if you had multiple possible block formats under the same PEN, you could make the payload type be an enum, with one variant for each block format. But even then, if you had multiple PENs to consider, you'd end up writing something like:
> 
> ```rust
> match read_block.pen {
>     MY_CUSTOM_PEN => {
>         let custom_payload = read_block.payload.as_any().downcast_ref::<MyCustomEnum>()?;
>         match custom_payload {
>             ...
>         }
>     }
>     MY_OTHER_PEN => {
>         let other_payload = read_block.payload.as_any().downcast_ref::<MyOtherEnum>()?;
>         match other_payload {
>             ...
>         }
>     }
> }
> ```
> 
> At which point, you've just reimplemented the dispatch mechanism anyway - so you might as well have just matched on the PEN and then called your parser functions directly, rather than registering them with the library.
> 
> I think my feeling is really that the read dispatch should be left to the user.

I get your point. It's true that with multiple PENs you would end up having to write a similar block to downcast them. I was just trying to homogenize reading and writing APIs because having them use different types seems a bit weird, but the complexity cost is probably too much.

I think your 4-type proposal is a good compromise.

courvoif commented:
@martinling I reviewed your proposal this day and I think we can simplify it and support @amorenoz usecase (wanting to create all the Blocks first).

The main thing I've done is to change the definition of the `CustomNonCopiable` trait to add a new associated type named State:
```rust
/* ----- trait CustomNonCopiable ----- */

/// Common interface for non-copiable custom block payloads
pub trait CustomNonCopiable<'a> {
    /// Private Enterprise Number of the entity which defined this payload format.
    const PEN: u32;

    /// External state that can be needed in order to parse/write the block.
    type State;

    /// Try to parse this payload from a slice.
    fn from_slice(state: &Self::State, slice: &'a [u8]) -> Result<Self, PcapError>
    where
        Self: Sized;

    /// Write this payload into a writer.
    fn write_to<W: Write>(&self, state: &Self::State, writer: &mut W) -> Result<usize, PcapError>;

    /// Convert this block in a non-copiable [`CustomBlock`]
    fn into_custom_block(self, state: &Self::State) -> Result<CustomBlock<'a, false>, PcapError>
    where
        Self: std::marker::Sized,
    {
        let mut data = Vec::new();
        self.write_to(state, &mut data)?;

        Ok(CustomBlock { pen: Self::PEN, payload: Cow::Owned(data) })
    }
}
```

The whole goal of the custom blocks is to be able to write new blocks that can depend on anything.
Therefore, I think the custom blocks must not have a dependency on `PcapNgState` because this state is about the "known" blocks and it would lessen the variety of blocks someone can write.
The direct consequence is that the user will be responsible for updating the state of his custom blocks.
And this change allowed me to revert the modifications done on the `ReadBuffer` and the new `Parser` trait because we don't need to get the state from `PcagNgWriter` anymore. 


Regarding @amorenoz usecase I think we can tackle it like this:
```rust
    let original_payload = MyCustomPayload { magic_number: 0xDEADBEEFCAFED00D };

    let mut buffer = Vec::new();
    let mut pcapng_writer = PcapNgWriter::new(&mut buffer)
        .expect("Failed to create writer");

    let block_to_write = original_payload.clone().into_custom_block().into_block();
    pcapng_writer
        .write_block(&block_to_write)
        .expect("Failed to write custom block");
```

In this exemple, we could create a `Vec<Block>` and pushing to it any kind of blocks including custom ones.

It's still a WIP but you can see the changes on the branch [courvoif/custom_block_api](https://github.com/courvoif/pcap-file/tree/courvoif/custom_block_api)

martinling commented:
Hi @courvoif, thanks for reviewing.

I agree that we need to support the case where encoding/decoding a custom block requires more context than we have in the standard `PcapNgState`, i.e. it needs to be possible to _add_ "external" state.

But what about the case where the format depends on the "internal" `PcapNgState`?

In many cases it will make sense for a custom block format to re-use formats that are already defined in the PcapNg standard, i.e. by specifying that the endianness and timestamp format of a custom block will follow those of the rest of the file. This allows programs to use the same code for reading/writing both EPBs and custom blocks.

This is why I made the `Parser` changes, because for such a block format, it needs to be possible to have the current `PcapNgState` accessible when interpreting a block.

So I think we need to _add_ the support for external state rather than have it _replace_ the support for internal state.

courvoif commented:
If a user want to track the same state that we use, he can use its own instance of a PcapNgState:
```rust
fn user_pcapng_state() {
    let data = [0_u8; 10];
    let ref_mut = &mut &data[..];

    let mut reader = PcapNgReader::new(ref_mut).unwrap();
    let mut state = PcapNgState::default();

    while let Some(blk) = reader.next_block() {
        let blk = blk.unwrap();
        state.update_from_block(&blk).unwrap();

        // Do something with the state
        dbg!(&state);
    }
}
````


If you think that it is really important for a user to be able to reuse the same state that we keep I may have a solution.
I don't like it but it has the advantage of being really simple:
```rust
    /// Returns the next [`Block`].
    pub fn next_block(&mut self) -> Option<Result<(Block, &PcapNgState), PcapError>> {
        match self.reader.has_data_left() {
            Ok(has_data) => {
                if has_data {
                    // # SAFETY
                    // Block must NOT contain a mutable reference to the state.
                    // Keep the annotations to be sure that only the lifetime is trnasmuted.
                    let res: Result<Block<'_>, PcapError> = self.reader.parse_with(|src| self.parser.next_block(src));
                    let res: Result<Block<'_>, PcapError> = unsafe {std::mem::transmute(res)};

                    let state = self.parser.state();

                    Some(res.map(|blk| (blk, state)))
                }
                else {
                    None
                }
            },
            Err(e) => Some(Err(PcapError::IoError(e))),
        }
    }
```

martinling commented:
> If a user want to track the same state that we use, he can use its own instance of a PcapNgState:

Yes, this is possible but it seems rather inelegant since it means all the work of maintaining the state is duplicated. If I found myself needing to do it like this, I would quickly be writing a PR to make it unnecessary!

> If you think that it is really important for a user to be able to reuse the same state that we keep, I may have a solution. I don't like it but it has the advantage of being really simple:
> 
> ```rust
>     /// Returns the next [`Block`].
>     pub fn next_block(&mut self) -> Option<Result<(Block, &PcapNgState), PcapError>> {
>         match self.reader.has_data_left() {
>             Ok(has_data) => {
>                 if has_data {
>                     // # SAFETY
>                     // Block must NOT contain a mutable reference to the state.
>                     // Keep the annotations to be sure that only the lifetime is trnasmuted.
>                     let res: Result<Block<'_>, PcapError> = self.reader.parse_with(|src| self.parser.next_block(src));
>                     let res: Result<Block<'_>, PcapError> = unsafe {std::mem::transmute(res)};
> 
>                     let state = self.parser.state();
> 
>                     Some(res.map(|blk| (blk, state)))
>                 }
>                 else {
>                     None
>                 }
>             },
>             Err(e) => Some(Err(PcapError::IoError(e))),
>         }
>     }
> ```

That accomplishes the same thing as my `next_block_and_state` implementation. So if you prefer that version of the innards of that method to my `Parser` refactoring then that's fine with me. Personally I was quite pleased with that solution.

martinling commented:
I think we're more or less on the same page @courvoif, but let me go through some details.

Your new API addresses @amorenoz's use case, constructing copiable custom `Blocks` in advance before passing them to a `PcapNgWriter`.

For my current use case in Packetry, I would be implementing your `CustomNonCopiable` with `State = PcapNgState`. To do that I would need a few changes from my branch that are currently missing from yours:

1. Your new implementation of `PcapNgReader::next_block_and_state`, as discussed above.
2. [Add PcapNgWriter::state() method](https://github.com/courvoif/pcap-file/pull/47/commits/87d8148800ebba0e0c1949b681270be82199cbe5).
3. [Add methods to get section and interfaces from PcapNgState](https://github.com/courvoif/pcap-file/pull/47/commits/40769e573281b04591e8e3b04f138456391217d5).
4. [Make encode_timestamp and decode_timestamp methods public](https://github.com/courvoif/pcap-file/pull/47/commits/8c5cecc8b76bf308ecc0d46d54dcf62769198be3).

Changes 1 and 2 are what enable me to re-use the internal `PcapNgState`. Changes 3 and 4 are necessary to make that object usable, even if I maintain my own.

I was wondering about the use case where some future custom block format depends on both internal and external state. But I think this could be done with `State = (&PcapNgState, &ExtState)`, and just constructing the necessary tuple when needed.

I still have one issue with your new API though:

What should a `from_slice` implementation do when it sees a format it does not recognise? One organisation may define many formats under the same PEN, distinguished by its own rules. Yet a program may only know about one or more of them.

This is why my version has `from_slice` and `interpret` return `Result<Option<T>, PcapError>` rather than `Result<T, PcapError>`. It would be possible to add a `PcapError` variant like `UnrecognisedFormat`, but fundamentally this is _not an error_, and shouldn't be treated as one. I think we should stick with the `Option` there.

martinling commented:
I've now rewritten this branch following the above discussion.

The API is as per @courvoif's proposal except that `from_slice` and `interpret` return `Result<Option<T>, PcapError>`.

I've added a test of the non-copiable use case for a format that includes state-dependent timestamp encoding.

In my use case, usage is a bit more verbose than before, because it's necessary to match on the endianness of the current `PcapNgState` in both the `from_slice` and `write_to` implementations, rather than having them generic over `ByteOrder`. 

I think that's reasonable though - it avoids putting that specific aspect of state into the trait definition, and it makes the intent to either use a fixed endianness, or follow that of the current section, very explicit in the code.

martinling commented:
Rewritten again after noticing a couple more API issues:

1. `into_custom_block` needs to return a `Result`, since the underlying `write_to` implementation is fallible. This tripped me up when I wrote a block with an invalid interface ID and called `encode_timestamp`. An error was being raised, but was thrown away by @courvoif's implementation of `into_custom_block`.

2. There's no point having `write_to` return a `usize` with the number of bytes written, since `into_custom_block` doesn't use this number. It gets the payload size automatically from the size of the `Vec` that it calls `write_to` on. Whcih is a much better plan anyway, since it's _really_ easy for the user to accidentally return the wrong size there.

amorenoz commented:
One issue I see here is how to deal with Endianness in `CustomCopiable` blocks.
We cannot assume the same Endianness as the machine we're running on and we don't have `PcapNgState.section().endianness` as `CustomNonCopiable`.

courvoif commented on src/pcapng/blocks/custom.rs:
```diff
@@ -0,0 +1,131 @@
+}
+
+/// Common interface for copiable custom block payloads
+pub trait CustomCopiable<'a> {
```
With `from_slice()` and `write_to() `the user may need to return custom errors. To support this use-case we could add an error type to both traits. What do you think ? 
In this case we would have to add a new variant to `PcapError` to handle the conversion between the user defined error and `PcapError` when converting a `CustomBlock` into a `Block`

martinling replied:
> With `from_slice()` and `write_to() `the user may need to return custom errors. To support this use-case we could add an error type to both traits. What do you think ?

Yes this is a good point, limiting the user to only the current `PcapError` variants is probably too restrictive here. I'm not sure if adding an error type to the trait is the best way forward though.

Unless we add _two_ error types, then both_ `from_slice` and `write_to` would need to return the same error type. Which may be annoying in practice, since `from_slice` may fail in complex format-dependent ways, but `write_to` is likely to be just calling into methods from `WriteBytesExt`.

> In this case we would have to add a new variant to `PcapError` to handle the conversion between the user defined error and `PcapError` when converting a `CustomBlock` into a `Block`

If we make that variant something like `CustomConversionError(Box<dyn std::error::Error>)`, then we wouldn't need the error type in the trait, the `from_slice` and `write_to` methods would be able to return it directly.

courvoif replied:
Everything else seems fine to me. Once we've settled on this topic I'll approve the PR.
